app/code/Magento/Inventory/Model/SourceRepository.php

```diff
@@ -0,0 +1,203 @@
+     */
+    private $collectionProcessor;
+
+    /**
+     * @var CollectionFactory
+     */
+    private $collectionFactory;
+
+    /**
+     * @var CarrierLinkCollectionFactory
+     */
+    private $carrierLinkCollectionFactory;
+
+    /**
+     * @var SourceSearchResultsInterfaceFactory
+     */
+    private $sourceSearchResultsFactory;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * SourceRepository constructor.
+     * @param ResourceSource $resourceSource
+     * @param ResourceSourceCarrierLink $resourceSourceCarrierLink
+     * @param SourceInterfaceFactory $sourceFactory
+     * @param CollectionProcessorInterface $collectionProcessor
+     * @param CollectionFactory $collectionFactory
+     * @param CarrierLinkCollectionFactory $carrierLinkCollectionFactory
+     * @param SourceSearchResultsInterfaceFactory $sourceSearchResultsFactory
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        ResourceSource $resourceSource,
+        ResourceSourceCarrierLink $resourceSourceCarrierLink,
+        SourceInterfaceFactory $sourceFactory,
+        CollectionProcessorInterface $collectionProcessor,
+        CollectionFactory $collectionFactory,
+        CarrierLinkCollectionFactory $carrierLinkCollectionFactory,
+        SourceSearchResultsInterfaceFactory $sourceSearchResultsFactory,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        LoggerInterface $logger
+    ) {
+        $this->resourceSource = $resourceSource;
+        $this->resourceSourceCarrierLink = $resourceSourceCarrierLink;
+        $this->sourceFactory = $sourceFactory;
+        $this->collectionProcessor = $collectionProcessor;
+        $this->collectionFactory = $collectionFactory;
+        $this->carrierLinkCollectionFactory = $carrierLinkCollectionFactory;
+        $this->sourceSearchResultsFactory = $sourceSearchResultsFactory;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function save(SourceInterface $source)
+    {
+        try {
+            $this->saveSource($source);
+            $this->saveSourceCarrierLinks($source);
+            return $source->getSourceId();
+        } catch (\Exception $exception) {
+            $this->logger->error($exception->getMessage());
+            throw new CouldNotSaveException(__('Could not save source'));
+        }
+    }
+
+    /**
+     * @param SourceInterface $source
+     */
+    private function saveSource(SourceInterface $source)
+    {
+        /** @var SourceInterface|AbstractModel $source */
+        $this->resourceSource->save($source);
+    }
+
+    /**
+     * @param SourceInterface $source
+     */
+    private function saveSourceCarrierLinks(SourceInterface $source)
```

Comment:
Is there a reason to extract this into a separate method? It seems to be used just once.